src/coin_test/data/datasets.py

```diff
@@ -0,0 +1,174 @@
+"""Define the Dataset classes."""
+
+from abc import ABCMeta, abstractmethod
+from collections import Counter
+from datetime import datetime
+from typing import Any
+
+import pandas as pd
+
+from .metadata import MetaData
+from .processors import Processor
+from ..util import AssetPair
+
+
+class DatasetMetaclass(type):
+    """Check class defines neccesary attributes.
+
+    Use to enforce the existence of specific attributes after object
+    instantiation. Unlike ABCMeta, does _not_ require the usage of @property,
+    and therefore works with setters and getter properly.
+    """
+
+    def __call__(cls, *args, **kwargs):  # noqa: ANN002,ANN003,ANN204
+        """Initialize class then check attributes."""
+        instance = super().__call__(*args, *kwargs)
+        if not hasattr(instance, "df"):
+            raise ValueError("`df` not defined in Dataset!")
+        return instance
```

Comment:
The `metaclass` usage here is slightly awkward - in general more esoteric Python like this seems like it should be avoided. However, it seems justified here as it allows us to require the `df` attribute while setting at runtime. This means we can use a setter that validates `df` before assigning it while also requiring all children to set it during initialization. 
(Note: This isn't possible with Python ABCMeta, so we make do with our own metaclass.)